crates/cext/src/circuit_library/qdrift.rs

```
@@ -0,0 +1,267 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::exit_codes::ExitCode;
+use crate::pointers::{const_ptr_as_ref, mut_ptr_as_ref};
+
+use num_complex::Complex64;
+
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::operations::Param;
+use qiskit_circuit_library::pauli_evolution::Instruction;
+use qiskit_circuit_library::pauli_evolution::sparse_term_evolution;
+use qiskit_quantum_info::sparse_observable::SparseObservable;
+
+use rand::distr::{Distribution, weighted::WeightedIndex};
+
+use std::ptr;
+
+/// Internal helper to extract real part of a complex number,
+/// panicking if imaginary part is non-zero.
+fn approx_real(z: &Complex64) -> Result<f64, ExitCode> {
+    if z.im.abs() > 1e-12 {
+        return Err(ExitCode::CInputError);
+    }
+    Ok(z.re)
+}
+
+/// This is an internal helper function to build the QDrift circuit.
+/// The same safety requirements as for `qk_circuit_library_qdrift` apply.
+fn qdrift_build_circuit(
+    obs: &SparseObservable,
+    reps: usize,
+    time: f64,
+) -> Result<CircuitData, ExitCode> {
+    let num_qubits = obs.num_qubits();
+    if reps == 0 || num_qubits == 0 || obs.coeffs().is_empty() {
+        return Err(ExitCode::CInputError);
+    }
+
+    // If the observable contains projectors, convert to Paulis
+    let pauli_obs = obs.as_paulis();
```

This potentially induces an exponential number of terms, if the user provides `n` projectors we get `2^n` Paulis. I would prefer to just return an error code if the operator contains projectors and let the user be explicit about potential conversion to Paulis.

---

Changed to return an error in 1a67ed5